### PR TITLE
update cma progressing condition

### PR DESCRIPTION
**Pull request by haoqing0110**

### CODE REVIEW

File: pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go [hunk 1]
```diff
@@ -72,15 +75,50 @@ func (d *managedClusterAddonConfigurationReconciler) reconcile(
...
-	for _, addon := range graph.addonToUpdate() {
+	nonplacementaddons, placementaddons := graph.addonToUpdateByPlacement()
+	// update all non placement addons
+	if err := d.updateAddonsConfig(ctx, nonplacementaddons); err != nil {
+		errs = append(errs, err...)
+	}
+
+	// update placement addons
+	for _, placementStrategy := range cma.Spec.InstallStrategy.Placements {
+		rollouttype := placementStrategy.RolloutStrategy.Type
```
Comment (qiujian16) [inline pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go]: should we just put RolloutStrategy into the graph, so we do not check cma here again

File: pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go [hunk 2]
```diff
@@ -72,15 +75,50 @@ func (d *managedClusterAddonConfigurationReconciler) reconcile(
...
+		case rollouttype == "" || rollouttype == addonv1alpha1.AddonRolloutStrategyUpdateAll:
+			if addons, exist := placementaddons[placementStrategy.PlacementRef]; exist {
+				if err := d.updateAddonsConfig(ctx, addons); err != nil {
+					errs = append(errs, err...)
+				}
+			}
+		case rollouttype == addonv1alpha1.AddonRolloutStrategyRollingUpdate:
+			placementNodes := graph.getPlacementNodes()
+			if installStrategyNode, exist := placementNodes[placementStrategy.PlacementRef]; exist {
+				length, err := parseMaxConcurrency(placementStrategy.RolloutStrategy.RollingUpdate.MaxConcurrency, len(installStrategyNode.clusters))
```
Comment (qiujian16) [inline pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go]: again I think this could be all done when building graph

**Review by qiujian16 — COMMENTED** [review]
I think we should maintain info in graph
1. the rolling strategy info for each placement node. 
2. the upgrade status for each addon node.

File: pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go (hunk 2 above)
Comment (haoqing0110) [reply pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go]: fixed.

File: pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go (hunk 1 above)
Comment (haoqing0110) [reply pkg/manager/controllers/addonconfiguration/addon_configuration_reconciler.go]: fixed

File: pkg/manager/controllers/addonconfiguration/graph.go [hunk 3]
```diff
@@ -17,20 +24,59 @@ type configurationGraph struct {
...
 // addonNode is node as a child of installStrategy node represting a mca
+// addonnode
 type addonNode struct {
 	desiredConfigs addonConfigMap
 	mca            *addonv1alpha1.ManagedClusterAddOn
+	// record if mca upgrade status
+	upgradeStatus int
 }
 
+const (
```
Comment (qiujian16) [inline pkg/manager/controllers/addonconfiguration/graph.go:51]: set it as a enum type?

File: pkg/manager/controllers/addonconfiguration/graph.go [hunk 4]
```diff
@@ -87,13 +134,31 @@ func (g *configurationGraph) addAddonNode(mca *addonv1alpha1.ManagedClusterAddOn
...
 		children:       map[string]*addonNode{},
 		clusters:       sets.New[string](clusters...),
 	}
 
+	// set max concurrency
+	if installStrategy.RolloutStrategy.Type == addonv1alpha1.AddonRolloutStrategyRollingUpdate {
+		if installStrategy.RolloutStrategy.RollingUpdate != nil {
+			node.maxConcurrency = installStrategy.RolloutStrategy.RollingUpdate.MaxConcurrency
+		} else {
+			node.maxConcurrency = intstr.FromString("25%")
```
Comment (qiujian16) [inline pkg/manager/controllers/addonconfiguration/graph.go]: should have a default variable for this

File: pkg/manager/controllers/addonconfiguration/graph.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/manager/controllers/addonconfiguration/graph.go:51]: fixed

File: pkg/manager/controllers/addonconfiguration/graph.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/manager/controllers/addonconfiguration/graph.go]: fixed

**Pull request merged**